File: AIペアプログラミング実践編_ソースコード/practical-java/session-5/Ans/src/main/java/com/example/inventory/service/OrderService.java
package com.example.inventory.service;

import com.example.inventory.dto.OrderInput;
import com.example.inventory.entity.Order;
import com.example.inventory.entity.OrderStatus;
import com.example.inventory.repository.OrderRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 発注サービス
 */
@Service
@Transactional
public class OrderService {

    private final OrderRepository orderRepository;
    private final InventoryService inventoryService;

    public OrderService(OrderRepository orderRepository, InventoryService inventoryService) {
        this.orderRepository = orderRepository;
        this.inventoryService = inventoryService;
    }

    /**
     * 発注を作成
     */
    public Order create(OrderInput input) {
        Order order = new Order();
        order.setProductId(input.getProductId());
        order.setQuantity(input.getQuantity());
        order.setStatus(OrderStatus.PENDING);
        order.setCreatedAt(LocalDateTime.now());
        return orderRepository.save(order);
    }

    /**
     * 発注を発注済みに更新
     */
    public Order order(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("発注が見つかりません"));

        if (order.getStatus() != OrderStatus.PENDING) {
            throw new IllegalStateException("発注中の発注のみ発注済みにできます");
        }

        order.setStatus(OrderStatus.ORDERED);
        order.setOrderedAt(LocalDateTime.now());
        return orderRepository.save(order);
    }

    /**
     * 入荷処理
     */
    public Order receive(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("発注が見つかりません"));

        if (order.getStatus() != OrderStatus.ORDERED) {
            throw new IllegalStateException("発注済みの発注のみ入荷処理できます");
        }

        // 在庫を更新
        inventoryService.updateQuantity(order.getProductId(), "IN", order.getQuantity());

        order.setStatus(OrderStatus.RECEIVED);
        order.setReceivedAt(LocalDateTime.now());
        return orderRepository.save(order);
    }

    /**
     * 発注をキャンセル
     */
    public Order cancel(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("発注が見つかりません"));

        if (order.getStatus() == OrderStatus.RECEIVED) {
            throw new IllegalStateException("入荷済みの発注はキャンセルできません");
        }

        order.setStatus(OrderStatus.CANCELLED);
        return orderRepository.save(order);
    }

    /**
     * 全発注を取得
     */
    public List<Order> findAll() {
        return orderRepository.findAll();
    }

    /**
     * ステータスで絞り込んで発注を取得
     */
    public List<Order> findByStatus(OrderStatus status) {
        return orderRepository.findByStatus(status);
    }
}

File: AIペアプログラミング実践編_ソースコード/practical-ts/session-5/Ans/src/services/orderService.ts
import type { Order, OrderInput, OrderStatus } from "../models"
import type { InventoryService } from "./inventoryService"
import { TransactionType } from "../models"

interface OrderRepository {
  save(input: OrderInput & { status: OrderStatus }): Promise<Order>
  findById(id: number): Promise<Order | null>
  findAll(): Promise<Order[]>
  update(id: number, input: { status: OrderStatus }): Promise<Order>
}

/**
 * 発注サービス
 * 発注に関するビジネスロジックを担当
 */
export class OrderService {
  constructor(
    private orderRepository: OrderRepository,
    private inventoryService: InventoryService,
  ) {}

  /**
   * 発注を作成
   */
  async create(input: OrderInput): Promise<Order> {
    // バリデーション
    if (!input.items || input.items.length === 0) {
      throw new Error("発注明細が必要です")
    }

    for (const item of input.items) {
      if (item.quantity <= 0) {
        throw new Error("数量は1以上で指定してください")
      }
    }

    // 発注を保存（ステータスは PENDING）
    return this.orderRepository.save({
      ...input,
      status: "PENDING",
    })
  }

  /**
   * 発注を確定
   */
  async confirm(orderId: number): Promise<Order> {
    const order = await this.orderRepository.findById(orderId)

    if (!order) {
      throw new Error("発注が見つかりません")
    }

    if (order.status !== "PENDING") {
      throw new Error("確定済みまたはキャンセル済みの発注は変更できません")
    }

    // 在庫を更新（入庫）
    for (const item of order.items) {
      await this.inventoryService.updateStock(
        item.productId,
        item.quantity,
        TransactionType.IN,
        `発注 #${orderId} 入荷`,
      )
    }

    // ステータスを更新
    return this.orderRepository.update(orderId, { status: "CONFIRMED" })
  }

  /**
   * 発注をキャンセル
   */
  async cancel(orderId: number): Promise<Order> {
    const order = await this.orderRepository.findById(orderId)

    if (!order) {
      throw new Error("発注が見つかりません")
    }

    if (order.status !== "PENDING") {
      throw new Error("確定済みまたはキャンセル済みの発注は変更できません")
    }

    return this.orderRepository.update(orderId, { status: "CANCELLED" })
  }

  /**
   * 発注一覧を取得
   */
  async findAll(): Promise<Order[]> {
    return this.orderRepository.findAll()
  }

  /**
   * 発注詳細を取得
   */
  async findById(id: number): Promise<Order | null> {
    return this.orderRepository.findById(id)
  }
}
